### PR TITLE
Mesh Adaptivity based on entropy of computed solution

### PR DESCRIPTION
**Pull request by maximilianmetti**
This PR contains a new feature that allows for mesh adaptivity based on the entropy of the computed solution. 

The benchmark problem in `benchmarks/pnp_refine_exact/` uses the new feature. Although, it's admittedly a silly example since the artificially chosen exact solution has non-zero entropy (due to the drift-diffusion source terms).  A new benchmark with more physical relevance should be put together to verify that the mesh refinement focuses on the regions of largest entropy.

### CODE REVIEW

**Comment by arthbous** [comment]
Which function apply the refinement ?

I see total charge that compute the charge but that's it.

**Comment by maximilianmetti** [comment]
I think there might be a bug 🐜... it seems to work well, but there may be a problem transferring the solution across the mesh after multiple refinements... I need to investigate. 

The refinement is that line to 225 of main.cpp. The code is also been re-factored to pull the Newton solve outside of the main function. We might consider doing that for all of our files in the future.

File: src/mesh_refiner.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+Mesh_Refiner::~Mesh_Refiner () {};
+//--------------------------------
+
+//--------------------------------
+std::shared_ptr<const dolfin::Mesh> Mesh_Refiner::get_mesh () {
+  return _mesh;
+};
+//--------------------------------
+std::shared_ptr<const dolfin::Mesh> Mesh_Refiner::multilevel_refinement (
+  std::vector<std::shared_ptr<const dolfin::Function>> entropy_potential_vector,
```
Comment (arthbous) [inline src/mesh_refiner.cpp:49]: In math what is entropy_potential_vector exactly ?

File: src/mesh_refiner.cpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  std::vector<std::shared_ptr<const dolfin::Function>> entropy_log_weight_vector
```
Comment (arthbous) [inline src/mesh_refiner.cpp:50]: In math what is entropy_log_weight_vector exactly ?

**Review by arthbous — COMMENTED** [review]
And Where do you refine the solutions ?

File: src/mesh_refiner.cpp (hunk 1 above)
Comment (maximilianmetti) [reply src/mesh_refiner.cpp:49]: it's a bad name for the term `\eta_i + q_i \phi`, which is used to compute the entropy:
`D_i exp(\eta_i) \nabla(\eta_i + q_i \phi)`.

File: src/mesh_refiner.cpp (hunk 2 above)
Comment (maximilianmetti) [reply src/mesh_refiner.cpp:50]: This is the term `D_i exp(\eta_i)` which is the multiplicative factor for the entropy mentioned above.

**Comment by maximilianmetti** [comment]
@arthbous I'd recommend looking at the [good-diode PR #75 ](https://github.com/thepnpsolver/modular-pnp/pull/75) instead of this one, as it has better mesh adaptivity.  That PR is up-to-date with my local branch.

I'm happy to close this PR so that we can merge PR #75 instead.

**Comment by arthbous** [comment]
So can we merge this guy ? Or should we merge the Diiode ?

**Comment by maximilianmetti** [comment]
Let's merge the diode instead

**Pull request closed without merging**